### ISSUE TITLE
On the Structural Integrity and Efficient Indexing of Sparse Trees

### ISSUE THREAD

**Issue by Alexander-Miller** [Alexander-Miller/treemacs#143]
Or how to greatly streamline the treemacs codebase while making it smaller, improve performance and make future changes that much easier to accomplish. This document shall serve as a guideline and a means to gather my thoughts before I jump head first into implementation (as if I didn't do that already).

## The Why?

Recent additions and the general proliferation of minor modes saw a bit of an explosion of caching layers in treemacs.

 1) There is one cache layer tracking which nodes are expanded under which parent, such that reopening a parent node will also reopen any previously open child nodes.
 2) There is another cache layer doing the same thing for tag nodes.
 3) Another cache keeps track of nodes' positions, greatly improving the performance of moving point to a given path over the previous `split search` approach (which is now used only as fallback).
 4) A further cache keeps track of directories being watched for changes. This happens even if `treemacs-filewatch-mode` is disabled, since treemacs must changes all these caches in reaction to fils being renamed or deleted.

Each cache needs to be tracked and accessed individually, adding to mental and technical overhead.

Additionally if treemacs is to switch to a multi-root layout (including non-filesystem roots like buffers or bookmarks) it would be well served by a data structure capable of tracking, updating and navigating through these roots.

### The Node Tree

The idea is to introduce a tree data structure which will shadow the view, collecting and updating nodes' status and metadata on each user interaction. The basic building block is the `node` structure, as defined by `cl-defstruct`. For the most basic functionality (single root, no filewatch) it is made of 4 fields: its key, parent, children, and position.

 - The key is a unique identifier for the node, assuring O(1) access to every node, no matter how deeply nested it is. For files this is their absolute path for tags their Tag Path, as already used in treemacs-tags.el.
 - The parent is the node below which the current node is expanded.
 - The children are a list of nodes expanded below the current node. The tree can therefore be traversed in any direction. The child/parent relationship also replaces the current open dir cache.
 - The position is a marker towards the node's position in the buffer, supplanting the current position cache, and allowing for O(1) movement to any visible node.

The tree structure will also simplify and increase the speed of actions that need to run on all generations of a node's children, such as updating or invalidating cache layers after the renaming or deletion of a file.

### The Node Index

Access to nodes inside the tree is handled via a single hash table, called the node index. Every single node is mapped to its key, assuring O(1) access and alleviating the need for complex tree traversal algorithms beyond a single generic routine to call a function on a node and all its children.

### The Troublesome Details (so far)

The presence of the open dirs cache makes it impossible for the node tree to be a *direct* mirror of the treemacs view. Closing a node must not lead to its removal from the index. If there exist other expanded nodes below the node being collapsed these nodes must be expanded again when their now closed parent is reopened. This puts a certain burden of complexity on even the basic on-expand/on-collapse routines mainly responsible for managing the node tree.

Closing a node must therefore merely sever it from its parent, keeping its relationships with its child generations intact (aside from invalidating their position data), while expanding a node must potentially restore this relation. This must also be taken into consideration when changing the root in an upwards or downwards direction.

The node tree is therefore not a single unit, but can instead resemble a collection of branches. Thus the word "sparse" in the title.

TL;DR: Make a tree, get O(1) everywhere.

**Comment by Alexander-Miller** [comment]
Nomenclature: since this whole new structure is shadowing the visible file tree I'll be calling it the shadow tree/node/index.

Sparse trees are out, they'll make efficiently applying code on a subset of the tree impossible. Closed subtrees will will simply receive a flag to mark them as such. This also allows for efficient refresh: registering a file change merely sets a flag on the containing directory. A refresh then merely entails recursively descending the shadow tree and only refreshing (really collapsing and re-expanding) marked directories. This way only the top-most changed directories are rebuilt (refresh-marked nodes below them will simply have their refresh flags reset).

**Comment by Alexander-Miller** [comment]
Done.
